modules/cas_cache/layer_cache_management.c

```diff
@@ -448,39 +465,44 @@ static void _cache_mngt_cache_stop_rollback_complete(ocf_cache_t cache,
 	else
 		BUG_ON(error);
 
-	INIT_WORK(&ctx->work, cache_start_rollback);
-	schedule_work(&ctx->work);
+	BUG_ON(!wake_up_process(ctx->rollback_thread));
 }
 
 static void _cache_mngt_cache_stop_complete(ocf_cache_t cache, void *priv,
 		int error)
 {
-	struct _cache_mngt_async_context *context = priv;
-	int result = exit_instance_finish(cache, error);
-
-	result = _cache_mngt_async_callee_set_result(context, result);
+	struct _cache_mngt_stop_context *context = priv;
+	context->error = error;
 
-	if (result == -KCAS_ERR_WAITING_INTERRUPTED)
-		kfree(context);
+	BUG_ON(!wake_up_process(context->finish_thread));
 }
 
 static int _cache_mngt_cache_stop_sync(ocf_cache_t cache)
 {
-	struct _cache_mngt_async_context *context;
+	struct _cache_mngt_stop_context *context;
 	int result = 0;
 
 	context = env_malloc(sizeof(*context), GFP_KERNEL);
 	if (!context)
 		return -ENOMEM;
 
-	_cache_mngt_async_context_init(context);
+	context->finish_thread = kthread_create(exit_instance_finish, context,
```

Comment:
I think this should also be moved somewhere to init phase.  This function is used in initialization error handling path (see err_classifier label) - and in case of _cache_mngt_cache_stop_sync error we leak resources.

Looks to me that the entire _cache_mngt_stop_context  together with context->finish_thread should be allocated on init phase and stored somewhere in cache priv.